### DISCUSSION TITLE
Pre-made custom meta fields on donation forms

### DISCUSSION THREAD

**Discussion by chamb55** [impress-org/givewp#7190] [Developer Q&A]
Im trying to create pre-made custom meta fields on donation forms when they are created, without the manual addition of custom fields using Visual Form Builder. Specifically, I aim to include an **image upload button.** To address this, I've been exploring solutions using the API and the givewp-snippet-library.

I need to add an "image upload button " to the default form programmatically. I tried this two methods.

**1. First Method**
 
```
//Add custom meta data to donation form //
function give_add_donor_company_fields( $form_id ) {
<div id="give-company-wrap" class="form-row form-row-wide">`

<label class="give-label" for="give-donor-logo splabel">
<?php esc_html_e( 'Company Logo', 'give' ); ?>
</label>
<input
class="give-input"
type="file"
name="give_donor_logo"
id="give-donor-logo"
required=""
aria-required="true"
accept="image/jpeg, image/png, image/gif, image/jpg"
>

</div>
<?php
}
add_action( 'give_donation_form_after_email', 'give_add_donor_company_fields' );


/** Save custom meta data  to meta.*/

function give_save_donor_company_details( $donation_id ) {
$donor_id = give_get_payment_donor_id( $donation_id );

// Company Logo
$uploaded_logo = $_FILES['give_donor_logo'];

if ( isset( $uploaded_logo['tmp_name'] ) && ! empty( $uploaded_logo['tmp_name'] ) ) {
$upload_overrides = array( 'test_form' => false );
$movefile = wp_handle_upload( $uploaded_logo, $upload_overrides );

if ( $movefile && ! isset( $movefile['error'] ) ) {
Give()->donor_meta->update_meta( $donor_id, '_give_donor_logo', $movefile['url'] );
} 
else {
// Display an error if the file upload fails

give_set_error( 'invalid_logo_upload', __( 'Error uploading company logo. Please try again.', 'give' ) );

}
}
}
add_action( 'give_insert_payment', 'give_save_donor_company_details', 10 );

// This Method is not working It not updating give_donationmeta Table 
```
 

**2. Second Method**

```
add_action( 'give_fields_after_donation_amount', function( $collection ) {
$collection->append(
give_field( 'file', 'give_donor_logo' )
->showInReceipt()
->label( __('Upload Image') )
->storeAsDonorMeta()
->required()
->helpText( __( 'Please upload an image for your donation.' ) )
);
});


```

The second method, using API, is working. However, I need more information about the "File" method and its related attributes. I need to limit the file size to 1MB and allow file formats with .jpeg, .png, and .gif file extensions. Currently, the meta field value is saved in the database as a meta value, but I need to save it as a file directory path. Is there any way to achieve this?

Referrer links
donor Meta Method : [(https://github.com/impress-org/givewp-snippet-library/blob/master/donors/add-donor-meta.php)](https://github.com/impress-org/givewp-snippet-library/blob/master/donors/add-donor-meta.php)

second Method : Donor meta using the Fields API: [(https://givewp.com/documentation/developers/how-to-create-custom-form-fields/)](https://givewp.com/documentation/developers/how-to-create-custom-form-fields/)

**Comment by jonwaldstein** [comment]
Hey @chamb55 

Your second method is the recommended way.

The [file field](https://github.com/impress-org/givewp/blob/b3b3135b730be331e7c1e8addbf021de59d7f54b/src/Framework/FieldsAPI/File.php) comes with additional validation methods:
`->maxUploadSize(int $size)`
`->allowedMimeTypes(array $types)`

However, by default this file field is processed as a WordPress attachment and the meta refers to the attachment ID.
